### PR TITLE
Create separate interfaces for streams

### PR DESCRIPTION
**Pull request by shadowhand**
Clearly differentiate between the different sources for streams.

Refs #2
Refs #4

### CODE REVIEW

**Comment by harikt** [comment]
Does't `StreamFromFileFactoryInterface`, `StreamFromResourceFactoryInterface` looks nice than `StreamFileFactoryInterface` or `StreamResourceFactoryInterface`  etc ? 

Alternative thought is remove `StreamFrom` entirely and just have `FileFactoryInterface`, `ResourceFactoryInterface`

I understand first one is a bit lengthy. Thoughts ?

**Comment by shadowhand** [comment]
Having "from" here implies (to me) that any underlying resource would be a copy, which is not true in the case of resource or file streams.

**Comment by harikt** [comment]
So how does `createStreamFrom*` implies?

**Comment by shadowhand** [comment]
Classes are nouns, methods are verbs.

**Comment by Nyholm** [comment]
👎  See my comment in https://github.com/http-interop/http-factory/issues/4#issuecomment-241337277

File: src/StreamResourceFactoryInterface.php [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    /**
+     * Create a new stream from a resource.
+     *
+     * The resource MUST be readable and SHOULD be seekable. It MAY be writable.
+     *
+     * @param resource $resource
+     *
+     * @return StreamInterface
+     */
+    public function createStreamFromResource($resource);
```
Comment (mindplay-dk) [inline src/StreamResourceFactoryInterface.php:18]: we can use a static type-hint here, can't we?

Comment (shadowhand) [inline src/StreamResourceFactoryInterface.php:18]: There is no [type hint for resources](http://php.net/manual/en/functions.arguments.php#functions.arguments.type-declaration).

**Comment by Nyholm** [comment]
@mindplay-dk wrote on https://github.com/http-interop/http-factory/pull/15#issuecomment-241494576, I thought I move his comment here to keep the discussion in one place: 

Please consider the number of times it says "if" in the description of this method [StreamFactory::createStream].

Every "if" in that description implies an "if" in the implementation, so that's a lot of "ifs", and a lot of explanations and code required to untangle what are essentially different factory-types weaved into one.

This interface is several factories overloaded behind a single API.

In order for implementers to meet those requirements, they will have to write essentially a formalized version of the if/else-ladder described in the comments.

I'm concerned it would be difficult to mock such a dependency for testing.

Is there a case where the client does not know what type of stream they're going to create, or from what?

Note that I do mean that as a question, I'm not being sarcastic - there may be such a case? If there is, that would justify the existence of an overloaded factory method. If there isn't, then it's not a single responsibility, and then, SRP...

**Comment by Nyholm** [comment]
> In order for implementers to meet those requirements, they will have to write essentially a formalized version of the if/else-ladder described in the comments.

That is how the current implementations are. See [Zend](https://github.com/zendframework/zend-diactoros/blob/1.3.5/src/Stream.php#L299-L326) and [Guzzle](https://github.com/guzzle/psr7/blob/1.3.1/src/functions.php#L78-L117). Ie, it is a design decision made by the community. 

> Is there a case where the client does not know what type of stream they're going to create, or from what?
> 
> Note that I do mean that as a question, I'm not being sarcastic - there may be such a case? If there is, that would justify the existence of an overloaded factory method. If there isn't, then it's not a single responsibility, and then, SRP...

Good question! Yes, I believe there is such use case. Consider the mail client again. A mail client may have a method like this: 

``` php

/**
 * Send an email
 * @param string $to
 * @param string $subject
 * @param string|resource $body A string body or a resource of a file containing HTML
 **/
public function send($to, $subject, $body) {
  $httpclient = new HttpClient();
  $request = new Request('POST', 'http://api.com');
  $request = $request->withHeader('X-To', $to);
  $request = $request->withHeader('X-Subject', $subject);
  $request = $request->withHeader('X-Api-Key', $this->apiKey);
  $request = $request->body((new StreamFactory)->createStream($body));

  $httpclient->sendRequest($request);
}
```

Of course one could argue that the mail client should validate their input. One might even suggest one should do `MailClient::sendMailFromFile` and `MailClient::sendMailFromString`. The question is really, where should one put all the if-statements?

My argument is still: Since the if-statements are already in Guzzle and Zend AND since hiding that logic in the factory implementation make the factories easier to use we should make this interface simple as `StreamFactory::createStream`.

**Comment by mindplay-dk** [comment]
> > In order for implementers to meet those requirements, they will have to write essentially a formalized version of the if/else-ladder described in the comments.
> 
> That is how the current implementations are. See Zend and Guzzle. Ie, it is a design decision made by the community.

Diactoros has four different types of stream implementations, and Guzzle has somewhere around [15 different types](https://github.com/guzzle/streams/tree/master/src) of streams.

As for the [Stream type in Diactoros](https://github.com/zendframework/zend-diactoros/blob/master/src/Stream.php), this appears to be a resource-based stream-type. The constructor accepts `string|resource`, and the factory behavior is built into the class [here](https://github.com/zendframework/zend-diactoros/blob/master/src/Stream.php#L299), where it takes a `string` and opens a `resource` for you; the end result is a resource-based stream either way.

Guzzle does something similar, but uses a [static factory method](https://github.com/guzzle/streams/blob/master/src/Stream.php#L45) and supports `string|resource` as well as `StreamInterface`, a capability that Diactoros doesn't have.

Do we ignore the extra capabilities of Guzzle, or do we burden Zend with adding this extra capability?

**Comment by shadowhand** [comment]
It seems pretty clear to me that the right answer from a user standpoint is to type hint against `string|resource`. Guzzle, Diactoros, and [Slim does not](https://github.com/slimphp/Slim/blob/1dedd2592c4f7e088ee3a87a7936f5df71cb7593/Slim/Http/Stream.php#L92-L96) and adding such functionality is easy:

``` php
$resource = fopen('php://temp', 'w+');
fwrite($resource, $content);

return new Stream($resource);
```

As for file streams, I do not think that this package needs to support that functionality, as passing a file is a resource is quite easy:

``` php
$stream = $factory->createStream(fopen($path, 'r'));
```

In terms of user ergonomics, `string|resource` is ideal. For the top 3 `http-message` implementations, 2/3 will have a very simple implementation, and the other will be able to easily compensate.

**Comment by mindplay-dk** [comment]
> In terms of user ergonomics, string|resource is ideal. For the top 3 http-message implementations, 2/3 will have a very simple implementation, and the other will be able to easily compensate.

OK, sold :-)

**Pull request closed without merging**